openedx/core/djangoapps/call_stack_manager/core.py

```diff
@@ -0,0 +1,163 @@
+"""
+Get call stacks of Model Class
+in three cases-
+1. QuerySet API
+2. save()
+3. delete()
+
+classes:
+CallStackManager -  stores all stacks in global dictionary and logs
+CallStackMixin - used for Model save(), and delete() method
+
+Functions:
+capture_call_stack - global function used to store call stack
+
+Decorators:
+donottrack - mainly for the places where we know the calls. This decorator will let us not to track in specified cases
+
+How to use-
+1. Import following in the file where class to be tracked resides
+    from openedx.core.lib.call_stack_manager import CallStackManager, CallStackMixin
+2. Override objects of default manager by writing following in any model class which you want to track-
+    objects = CallStackManager()
+3. For tracking Save and Delete events-
+    Use mixin called "CallStackMixin"
+    For ex.
+        class StudentModule(CallStackMixin, models.Model):
+
+Note -
+1.Format for STACK_BOOK
+{
+    "modelclass1":
+        [[(frame 1),(frame 2)],
+         [(frame 11),(frame21)]]
+    "modelclass2":
+        [[(frame 3),(frame 4)],
+         [(frame 6),(frame 5)]]
+
+}
+where frame is a tuple of
+(file path, Line Number, Context)
+"""
+
+import logging
+import traceback
+import re
+import collections
+from django.db.models import Manager
+
+log = logging.getLogger(__name__)
+
+# Module Level variables
+# dictionary which stores call stacks.
+# { "ModelClasses" : [ListOfFrames]}
+# Frames - ('FilePath','LineNumber','Context')
+# ex. {"<class 'courseware.models.StudentModule'>" : [[(file,line number,context),(---,---,---)],
+#                                                     [(file,line number,context),(---,---,---)]]}
+STACK_BOOK = collections.defaultdict(list)
+
+# filter to trickle down call stacks
+EXCLUDE = ['^.*python2.7.*$', '^.*<exec_function>.*$', '^.*exec_code_object.*$']
+REGULAR_EXPS = [re.compile(x) for x in EXCLUDE]
+
+# Variable which decides whether to track calls in the function or not. Do it by default.
+TRACK_FLAG = True
+
+# List keeping track of Model classes not be tracked for special cases
+# usually cases where we know that the function is calling Model classes.
+HALT_TRACKING = []
+
+
+def capture_call_stack(current_model):
```

Comment:
A side-benefit of passing `current_model` in as a class, rather than a string here, is that in tests you can assert on the class object, rather than on a string.